README.md
# clj-dl4j-examples

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2023 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2023-03-03
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2023-03-03
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://sourcehost.site/your-name/clj-dl4j-examples/compare/0.1.1...HEAD
[0.1.1]: https://sourcehost.site/your-name/clj-dl4j-examples/compare/0.1.0...0.1.1


test/clj_dl4j_examples/core_test.clj
(ns clj-dl4j-examples.core-test
  (:require [clojure.test :refer :all]
            [clj-dl4j-examples.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


project.clj
(defproject io.github.jeremylcfr/clj-dl4j-examples "0.1.0-SNAPSHOT"
  :description "FIXME: write description"
  :url "http://example.com/FIXME"
  :license {:name "EPL-2.0 OR GPL-2.0-or-later WITH Classpath-exception-2.0"
            :url "https://www.eclipse.org/legal/epl-2.0/"}
  :dependencies [[org.clojure/clojure "1.11.1"]
                 [org.clojure/algo.generic "0.1.3"]
                 [org.jfree/jfreechart "1.5.4"]
                 ;; [org.nd4j/nd4j-cuda-11.6-platform "1.0.0-M2.1"]
                 [io.github.jeremylcfr/clj-nd4j "0.1.0-SNAPSHOT"]
                 [io.github.jeremylcfr/clj-datavec "0.1.0-SNAPSHOT"]
                 [io.github.jeremylcfr/clj-dl4j "0.1.0-SNAPSHOT"]]
  :source-paths ["src/clojure"]
:java-source-paths ["src/java"]
  :repl-options {:init-ns clj-dl4j-examples.core})


src/clojure/clj_dl4j_examples/examples/xor.clj
(ns clj-dl4j-examples.examples.xor
  (:require [clj-dl4j.core :as core]
            [clj-nd4j.dataset :as dataset]
            [clj-nd4j.ndarray :as nd4j]
            [clj-dl4j.datasets :as ddataset]
            [clj-java-commons.coerce :refer [->clj]])
  (:import [org.deeplearning4j.nn.multilayer MultiLayerNetwork]
           [org.deeplearning4j.nn.conf MultiLayerConfiguration]
           [org.nd4j.linalg.api.ndarray INDArray]
           [org.nd4j.linalg.dataset DataSet]))

(def xor-training-data
  (dataset/->dataset
    ;; Features [x y]
    [[0 0] [0 1] [1 0] [1 1]]
    ;; Labels (xor x y)
    ;; with [0 1] = false and [1 0] = true
    [[0 1] [1 0] [1 0] [0 1]]))

;; Datavec version from file
;; Vectorize data as behind
;; (def xor-training-data
;;   (ddataset/read-datasets :jackson
;;     {:type-hint      :regular
;;      :dataset-type   :classification
;;      :use-nd4j?      false
;;      :schema         [:x :y :xor]
;;      :batch-size     4
;;      :label-idx      2
;;      :max-labels     2}
;;     "data/xor.json"))

(def xor-network-configuration
  {:mini-batch false
   :seed 123
   :weights-updater {:type :sgd
                     :learning-rate 0.1}
   :bias-init 0
   :layers [{:type                   :dense
             :n-in                   2
             :n-out                  4
             :activation-fn          :sigmoid
             :weight-distribution    {:type   :uniform
                                      :lower  0
                                      :upper  1}}
            {:type                   :output
             :n-in                   4
             :n-out                  2
             :activation-fn          :softmax
             :loss-fn                :negative-log-likehood
             :weight-distribution    {:type   :uniform
                                      :lower  0
                                      :upper  1}}]
   :training-listener {:type         :score-iteration
                       :nb-iterations  100}})

(defn build-network
  ^MultiLayerNetwork
  []
  (core/multi-layer-network xor-network-configuration))

(def network (volatile! nil))

(defn build!
  []
  (vreset! network (build-network)))

(defn train!
  ([]
   (train! 1000))
  ([n]
   (if-let [network* @network]
     (dotimes [_ n] (.fit ^MultiLayerNetwork network* ^DataSet xor-training-data))
     (do (build!) (train! n)))))

(defn translate!
  [^INDArray a]
  (let [result (->clj a)]
    (println (str "Score for XOR({FALSE ; FALSE}) ==================> " (/ (Math/round ^double (* 100000 (get-in result [0 0]))) 1000.0) "% TRUE"))
    (println (str "Score for XOR({ TRUE ; FALSE}) ==================> " (/ (Math/round ^double (* 100000 (get-in result [1 0]))) 1000.0) "% TRUE"))
    (println (str "Score for XOR({FALSE ;  TRUE}) ==================> " (/ (Math/round ^double (* 100000 (get-in result [2 0]))) 1000.0) "% TRUE"))
    (println (str "Score for XOR({ TRUE ;  TRUE}) ==================> " (/ (Math/round ^double (* 100000 (get-in result [3 0]))) 1000.0) "% TRUE"))))

(defn show!
  ([]
   (show! (.getFeatures ^DataSet xor-training-data)))
  ([data]
    (if-let [network* @network]
     (translate! (.output ^MultiLayerNetwork network* ^INDArray (nd4j/nd-array :matrix data)))
     (do (build!) (show! data)))))


src/clojure/clj_dl4j_examples/examples/mnistanomaly.clj
(ns clj-dl4j-examples.examples.mnistanomaly
  (:require [clojure.algo.generic.functor :refer [fmap]]
            [clj-nd4j.ndarray :as nd4j]
            [clj-nd4j.dataset :as dt]
            [clj-dl4j.core :as core]
            [clj-dl4j.resources.datasets :as dts])
  (:import [clj_dl4j_examples.examples.mnistanomaly MNISTVisualizer]

           ;; Wrapped/hinted
           [org.deeplearning4j.nn.multilayer MultiLayerNetwork])
  (:refer-clojure :exclude [run!]))

(def data (volatile! nil))

(def mnistanomaly-network-configuration
  {:seed 12345
   :weight-init-method :xavier
   :weights-updater {:type :adagrad
                     :learning-rate 0.05}
   :activation-fn :relu
   :l2-weights 0.0001
   :layers [{:type :dense
             :n-in 784
             :n-out 250}
            {:type :dense
             :n-in 250
             :n-out 10}
            {:type :dense
             :n-in 10
             :n-out 250}
            {:type                   :output
             :n-in                   250
             :n-out                  784
             :loss-fn                :mse}]
   :training-listener {:type         :score-iteration
                       :nb-iterations  10}})

(defn build-network
  []
  (core/multi-layer-network mnistanomaly-network-configuration))

(defn ->dataset-iterator
  []
  (iterator-seq
    (dts/mnist-dataset-iterator 100 50000 false)))

(defn prepare-data
  []
  (let [iterator (->dataset-iterator)
        random (java.util.Random. 12345)]
    (reduce
      (fn [agg dataset]
        (let [{:keys [train test]} (dt/deep-split-test-and-train dataset 80 random)
              train-features (:features train)
              test-features (:features test)
              test-labels (nd4j/arg-max (:labels test) [1])]
          (-> (update agg :train-features conj train-features)
              (update     :test-features  conj test-features)
              (update     :test-labels    conj test-labels))))
      {:train-features []
       :test-features  []
       :test-labels    []} iterator)))

(defn prepare-data!
  []
  (vreset! data (prepare-data)))

(defn train!
  [^MultiLayerNetwork network nb-trains]
  (doseq [epoch (range nb-trains)]
    (doseq [feature (:train-features @data)]
      (.fit network feature feature))
    (println (str "Epoch " epoch " complete"))))

(defn fit-and-score-by-digits
  [^MultiLayerNetwork network]
  (let [{:keys [test-features test-labels]} @data]
    (reduce-kv
      (fn [super-agg idx feature]
        (let [label (nth test-labels idx)
              nrows (.rows feature)]
          (reduce
            (fn [agg i]
              (let [example (.getRow feature i true)
                    digit (long (.getDouble label i))
                    score (.score network (dt/->dataset example example))]
                (update agg digit conj {:score score :example example})))
            super-agg (range nrows))))
      {} test-features)))

(defn sort-results
  [results]
  (reduce-kv
    (fn [agg digit shard]
      (let [sorted (->> (sort-by :score shard)
                        (map :example))
            rsorted (reverse sorted)]
        (-> (update agg :best  concat (take 5 sorted))
            (update     :worst concat (take 5 rsorted)))))
    {} results))


;; ! Use this !
;; BEWARE : training on 30 periods is a pretty long task (and a pretty overkill one too)
(defn run!
  ([]
   (run! 30))
  ([nb-trains]
   (let [network (build-network)]
     (prepare-data!)
     (train! network nb-trains)
     (let [fitted (fit-and-score-by-digits network)
           {:keys [best worst]} (sort-results fitted)]
       (.visualize (MNISTVisualizer. 2.0 best "Best (Low Rec. Error)"))
       (.visualize (MNISTVisualizer. 2.0 worst "Worst (High Rec. Error)"))))))


src/java/clj_dl4j_examples/examples/mnistanomaly/MNISTVisualizer.java
/*******************************************************************************
 * Copyright (c) 2015-2019 Skymind, Inc.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * Adapted from nested class :
 * https://github.com/eclipse/deeplearning4j-examples/blob/master/dl4j-examples/src/main/java/org/deeplearning4j/examples/feedforward/anomalydetection/MNISTAnomalyExample.java
 ******************************************************************************/

package clj_dl4j_examples.examples.mnistanomaly;

import org.nd4j.linalg.api.ndarray.INDArray;
import javax.swing.*;
import java.awt.*;
import java.awt.image.BufferedImage;
import java.util.*;
import java.util.List;


public class MNISTVisualizer {

  private double imageScale;
  private List<INDArray> digits;  //Digits (as row vectors), one per INDArray
  private String title;
  private int gridWidth;

  public MNISTVisualizer(double imageScale, List<INDArray> digits, String title ) {
    this(imageScale, digits, title, 5);
  }

  public MNISTVisualizer(double imageScale, List<INDArray> digits, String title, int gridWidth ) {
    this.imageScale = imageScale;
    this.digits = digits;
    this.title = title;
    this.gridWidth = gridWidth;
  }

  public void visualize(){
    JFrame frame = new JFrame();
    frame.setTitle(title);
    frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);

    JPanel panel = new JPanel();
    panel.setLayout(new GridLayout(0,gridWidth));

    List<JLabel> list = getComponents();
    for(JLabel image : list){
      panel.add(image);
    }

    frame.add(panel);
    frame.setVisible(true);
    frame.pack();
  }

  private List<JLabel> getComponents(){
    List<JLabel> images = new ArrayList<>();
    for( INDArray arr : digits ){
      BufferedImage bi = new BufferedImage(28,28,BufferedImage.TYPE_BYTE_GRAY);
      for( int i=0; i<784; i++ ){
        bi.getRaster().setSample(i % 28, i / 28, 0, (int)(255*arr.getDouble(i)));
      }
      ImageIcon orig = new ImageIcon(bi);
      Image imageScaled = orig.getImage().getScaledInstance((int)(imageScale*28),(int)(imageScale*28),Image.SCALE_REPLICATE);
      ImageIcon scaled = new ImageIcon(imageScaled);
      images.add(new JLabel(scaled));
    }
    return images;
  }
}


src/clojure/clj_dl4j_examples/examples/regression.clj
(ns clj-dl4j-examples.examples.regression
  (:require [clj-dl4j.core :as core]
            [clj-dl4j.datasets :as datasets]
            [clj-datavec.records.csv :as csv]
            [clj-datavec.records :as records]
            [clj-nd4j.ndarray :as nda]
            [clj-nd4j.dataset.normalization :as norm])

  (:import [org.deeplearning4j.eval RegressionEvaluation]

           [org.nd4j.linalg.api.ndarray INDArray]


           [org.jfree.chart ChartFactory ChartPanel JFreeChart]
           [org.jfree.chart.axis NumberAxis]
           [org.jfree.chart.plot XYPlot PlotOrientation]
           [org.jfree.data.xy XYSeries XYSeriesCollection]
           [org.jfree.ui RefineryUtilities])
  )

;; {:keys [mini-batch-size nb-possible-labels label-idx regression] :or {regression false} :as options}

;; https://github.com/deeplearning4j/deeplearning4j-examples/blob/master/oreilly-book-dl4j-examples/dl4j-examples/src/main/java/org/deeplearning4j/examples/recurrent/regression/SingleTimestepRegressionExample.java

(def dataset-options
  {:input-type :csv
   :to-skip 0
   :sep ";"
   :path-template {:train "data/regression/passengers_train_%d.csv"
                   :test  "data/regression/passengers_test_%d.csv"}
   :mini-batch-size 32
   :nb-possible-labels -1
   :label-idx 1
   :regression true
   :split {:type :numbered-file
           :min-idx 0
           :max-idx 0}})


;; num-to-skip
(defn build-dataset-iterator
  [type-fn]
  (datasets/->sequence-record-reader-dataset-iterator  (update dataset-options :path-template type-fn)))

(defn build-dataset
  [type-fn]
  (.next (build-dataset-iterator type-fn)))

;; maybe one map
(def normalizer-options
  {:type :min-max-scaler
   :spec {:min-range 0.0
          :max-range 1.0}
   :options {:fit-label? true}})

(defn build-normalizer
  []
  (let [{:keys [type spec options]} normalizer-options]
    (norm/normalizer-min-max-scaler spec options)))

(def network-configuration
  {:seed 140
   :weight-init-method :xavier
   :weights-updater {:type :nesterovs
                     :learning-rate 0.0015
                     :momentum 0.9}
   :layers [{:type           :graves-lstm
             :activation-fn  :tanh
             :n-in           1
             :n-out          10}
            {:type           :rnn-output
             :activation-fn  :identity
             :loss-fn        :mse
             :n-in           10
             :n-out          1}]
   :training-listener {:type           :score-iteration
                       :nb-iterations  20}})

(defn create-plot-series
  ^XYSeriesCollection
  ([features offset plot-name]
   (create-plot-series features offset plot-name (XYSeriesCollection.)))
  ([^INDArray features offset ^String plot-name ^XYSeriesCollection series-collection]
   (let [nb-rows (aget (nda/get-shape features) 2)
         series (XYSeries. plot-name)]
     (doseq [i (range nb-rows)]
       (.add ^XYSeries series ^int (int (+ offset i))))
     series-collection)))


(defn run!
  []
  (let [train-dataset (build-dataset :train)
        test-dataset  (build-dataset :test)

        normalizer (-> (build-normalizer)
                       (norm/fit-dataset train-dataset)
                       )
        _ (norm/transform! normalizer train-dataset)
        _ (norm/transform! normalizer test-dataset)

        ;; replace
        test-features (.getFeatures test-dataset)
        test-labels (.getLabels test-dataset)

        network (core/multi-layer-network network-configuration)]
    (dotimes [k 300]
      (.fit network train-dataset)
      (let [evaluation (RegressionEvaluation. 1)
            predicted (.output network test-features, false)]
        (.evalTimeSeries evaluation test-labels predicted)
        (println (.stats evaluation))))))












;; (def data (volatile! nil))

;; (def mnistanomaly-network-configuration
;;   {:seed 12345
;;    :weight-init-method :xavier
;;    :weights-updater {:type :adagrad
;;                      :learning-rate 0.05}
;;    :activation-fn :relu
;;    :l2-weights 0.0001
;;    :layers [{:type :dense
;;              :n-in 784
;;              :n-out 250}
;;             {:type :dense
;;              :n-in 250
;;              :n-out 10}
;;             {:type :dense
;;              :n-in 10
;;              :n-out 250}
;;             {:type                   :output
;;              :n-in                   250
;;              :n-out                  784
;;              :loss-fn                :mse}]
;;    :training-listener {:type         :score-iteration
;;                        :nb-iterations  10}})

;; (defn build-network
;;   []
;;   (core/multi-layer-network mnistanomaly-network-configuration))

;; (defn ->dataset-iterator
;;   []
;;   (iterator-seq
;;    (dts/mnist-dataset-iterator 100 50000 false)))

;; (defn prepare-data
;;   []
;;   (let [iterator (->dataset-iterator)
;;         random (java.util.Random. 12345)]
;;     (reduce
;;      (fn [agg dataset]
;;        (let [{:keys [train test]} (dt/deep-split-test-and-train dataset 80 random)
;;              train-features (:features train)
;;              test-features (:features test)
;;              test-labels (nd4j/arg-max (:labels test) [1])]
;;          (-> (update agg :train-features conj train-features)
;;              (update     :test-features  conj test-features)
;;              (update     :test-labels    conj test-labels))))
;;      {:train-features []
;;       :test-features  []
;;       :test-labels    []} iterator)))

;; (defn prepare-data!
;;   []
;;   (vreset! data (prepare-data)))

;; (defn train!
;;   [^MultiLayerNetwork network nb-trains]
;;   (doseq [epoch (range nb-trains)]
;;     (doseq [feature (:train-features @data)]
;;       (.fit network feature feature))
;;     (println (str "Epoch " epoch " complete"))))

;; (defn fit-and-score-by-digits
;;   [^MultiLayerNetwork network]
;;   (let [{:keys [test-features test-labels]} @data]
;;     (reduce-kv
;;      (fn [super-agg idx feature]
;;        (let [label (nth test-labels idx)
;;              nrows (.rows feature)]
;;          (reduce
;;           (fn [agg i]
;;             (let [example (.getRow feature i true)
;;                   digit (long (.getDouble label i))
;;                   score (.score network (dt/->dataset example example))]
;;               (update agg digit conj {:score score :example example})))
;;           super-agg (range nrows))))
;;      {} test-features)))

;; (defn sort-results
;;   [results]
;;   (reduce-kv
;;    (fn [agg digit shard]
;;      (let [sorted (->> (sort-by :score shard)
;;                        (map :example))
;;            rsorted (reverse sorted)]
;;        (-> (update agg :best  concat (take 5 sorted))
;;            (update     :worst concat (take 5 rsorted)))))
;;    {} results))


;; ;; ! Use this !
;; ;; BEWARE : training on 30 periods is a pretty long task (and a pretty overkill one too)
;; (defn run!
;;   ([]
;;    (run! 30))
;;   ([nb-trains]
;;    (let [network (build-network)]
;;      (prepare-data!)
;;      (train! network nb-trains)
;;      (let [fitted (fit-and-score-by-digits network)
;;            {:keys [best worst]} (sort-results fitted)]
;;        (.visualize (MNISTVisualizer. 2.0 best "Best (Low Rec. Error)"))
;;        (.visualize (MNISTVisualizer. 2.0 worst "Worst (High Rec. Error)"))))))
